examples/llava/gemma3-cli.cpp

```diff
@@ -224,6 +161,45 @@ static int generate_response(gemma3_context & ctx, common_sampler * smpl, int n_
     return 0;
 }
 
+static int eval_message(gemma3_context & ctx, common_chat_msg & msg, std::vector<std::string> & images_fname, bool add_bos = false) {
+    std::vector<mtmd_bitmap> bitmaps;
+
+    common_chat_templates_inputs tmpl_inputs;
+    tmpl_inputs.messages = {msg};
+    tmpl_inputs.add_generation_prompt = true;
+    tmpl_inputs.use_jinja = false; // jinja is buggy here
+    auto formatted_chat = common_chat_templates_apply(ctx.tmpls.get(), tmpl_inputs);
+    LOG_DBG("formatted_chat.prompt: %s\n", formatted_chat.prompt.c_str());
+
+    for (auto & fname : images_fname) {
+        mtmd_bitmap bitmap;
+        if (mtmd_helper_bitmap_init_from_file(fname.c_str(), bitmap)) {
+            LOG_ERR("Unable to load image %s\n", fname.c_str());
+            return 2; // image not found
+        }
+        bitmaps.push_back(std::move(bitmap));
+    }
+
+    mtmd_input_text text;
+    text.text          = formatted_chat.prompt;
+    text.add_special   = add_bos;
+    text.parse_special = true;
+    mtmd_input_chunks_ptr chunks(mtmd_tokenize(ctx.ctx_vision.get(), text, bitmaps));
+    if (chunks == nullptr) {
+        LOG_ERR("Unable to tokenize prompt\n");
+        return 1;
+    }
+
+    if (mtmd_helper_eval(ctx.ctx_vision.get(), ctx.lctx, chunks.get(), ctx.n_past, 0, ctx.n_batch)) {
+        LOG_ERR("Unable to eval prompt\n");
+        return 1;
+    }
+
```

Comment:
I think the main question will be how the interaction between the `libmtmd` and `libllama` will work. In the case of Gemma 3, we already have the necessary API so all is good and we don't need changes in `libllama` to make it work. But for other models, there will be changes needed to the API and this is where we have to carefully think it through and it might take more effort to provide the necessary support.

Reply:
In the short-term, my idea is to have `libllama` using input from `libmtmd`. In other words, `libllama` only need to knows about text token and embeddings, why `libmtmd` will handle image/audio/etc and convert them into token/embd.

This approach will make these 2 library completely separated for now. When be bring mtmd into libllama, we will mostly focus how to pass data between the 2 parts more efficiently (for ex, pass by tensor as @slaren suggested)

For now, some `mtmd_helper_*` functions will be added, which may use `libllama` under the hood. Indeed, this is because we don't yet have a notion similar to `common.h` specifically for mtmd, but it will be simple to add in the future.

Reply:
> But for other models, there will be changes needed to the API and this is where we have to carefully think it through and it might take more effort to provide the necessary support.

This is something I'll try to avoid at this stage, probably even using hooks/hacks like what I did in my CSM impl. Then when more models using it, we can make an API for it